### PR TITLE
fix: expose call graph creation errors

### PR DESCRIPTION
**Pull request by muscar**
- [ ] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

This does two things (I know, I know):

- it accepts a timeout from the client, and passes it to the call graph generator; and
- it catches call graph generator errors, and reports them to the client, *but* continues the dep work.

This is meant to improve the experience around reachable vulns by:
- giving control to the CLI on how long the call graph generator can run;
- running the user command (e.g. monitor, test) even if the call graph generation failed;
- allowing for better error messages is the call graph generation fails.

This should only be merged after https://github.com/snyk/java-call-graph-builder/pull/21 and https://github.com/snyk/snyk-cli-interface/pull/36, and after updating the dependencies to the appropriate versions.

#### What are the relevant tickets?

https://snyksec.atlassian.net/browse/FLOW-336

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -165,13 +157,23 @@ export function buildArgs(
...
 
-function buildErrorMessage(
-  error: Error,
-  mvnArgs: string[],
-  mavenCommand: string,
-): string {
-  if (error instanceof CallGraphError) {
-    return formatCallGraphError(error);
+async function getCallGraph(targetPath: string, timeout?: number): Promise<CallGraphResult> {
+  debug(`getting call graph from path ${targetPath}`);
```
Comment (darmalovan) [inline lib/index.ts:161]: Can we add the timeout to the output of `debug` ?

File: lib/index.ts [hunk 2]
```diff
@@ -9,8 +8,9 @@ import debugModule = require('debug');
 import { parseTree, parseVersions } from './parse-mvn';
 import * as subProcess from './sub-process';
 import { containsJar, createPomForJar, createPomForJars, isJar } from './jar';
-import { CallGraphError } from './errors/call-graph-error';
 import { formatCallGraphError, formatGenericPluginError } from './error-format';
+import { CallGraphResult, CallGraphError } from '@snyk/cli-interface/legacy/common';
+import { CallGraph } from '@snyk/cli-interface/legacy/common';
```
Comment (darmalovan) [inline lib/index.ts:13]: Can unify with line 12

File: lib/index.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ function debug(s: string) {
 export interface MavenOptions extends legacyPlugin.BaseInspectOptions {
   scanAllUnmanaged?: boolean;
   reachableVulns?: boolean;
+  callGraphBuilderTimeout?: number;
```
Comment (darmalovan) [inline lib/index.ts:32]: Will be nice to write here if it's in seconds?

File: lib/index.ts [hunk 4]
```diff
@@ -110,22 +111,13 @@ export async function inspect(
...
-          callGraphMetrics: javaCallGraphBuilder.runtimeMetrics(),
-        };
-        debug('got call graph successfully');
-      } catch (err) {
-        debug('call graph error: ' + err);
-        throw new CallGraphError(err.message, err);
-      }
+      callGraph = await getCallGraph(targetPath, options?.callGraphBuilderTimeout);
+      maybeCallGraphMetrics = {
+        callGraphMetrics: javaCallGraphBuilder.runtimeMetrics(),
```
Comment (darmalovan) [inline lib/index.ts:119]: Also will be nice to add the timeout to the metrics, this we will be able to see it in the logs

File: lib/index.ts [hunk 5]
```diff
@@ -29,6 +32,7 @@ function debug(s: string) {
 export interface MavenOptions extends legacyPlugin.BaseInspectOptions {
   scanAllUnmanaged?: boolean;
   reachableVulns?: boolean;
+  callGraphBuilderTimeout?: number; // milliseconds, as accepted by setTimeout
```
Comment (darmalovan) [inline lib/index.ts:35]: Maybe better to expose seconds? I would expect the users to have seconds? Will be nice to have it communicated in the name for less confusion

Comment (muscar) [reply lib/index.ts:35]: I think milliseconds might give better control. Maybe having the unit in the name would be clearer.

Comment (darmalovan) [reply lib/index.ts:35]: Gotcha, I think in the CLI it should be exposed to the user as seconds, like other tools that expose timeout as cli args

Comment (muscar) [reply lib/index.ts:35]: Agreed! I'll make the user facing option expressed in seconds

**Pull request merged**